designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,155 @@
+# Karpenter Drift Hash Versioning with CRD
+
+## Summary of Drift Hashing Mechanism
+
+Drift utilizes a hashing mechanism to identify when a `NodeClaim` has been drifted from its owning `NodePool`. Karpenter will hash fields that are within the `NodePool.Spec.Template`, excluding [special cases of drift](https://karpenter.sh/docs/concepts/disruption/#special-cases-on-drift) or [behavior fields](https://karpenter.sh/docs/concepts/disruption/#behavioral-fields). The `NodePool` will be annotated with the hash. When a `NodeClaim` is created, the `NodePool` hash will be annotated on the `NodeClaim`. More details can be found in the [drift documentation](https://karpenter.sh/docs/concepts/disruption/#drift). 
+
+When a `NodePool` is updated, a new hash will be annotated on the `NodePool`. The `NodeClaim` disruption controller will periodically check if the `NodePool` hash match to determine when a `NodeClaim` is drifted.
+
+## Background
+
+Karpenter's static drift mechanism works well when adding new fields to the CRD that don't have defaults, but breaks down if we add a field that has a set default, remove a field from the CRD, or set a new default for an existing field. If a cloud provider makes one of these changes to their CRDs, the static drift would change, causing all nodes to roll when the CRD is upgraded. This causes problems for cloud providers when maintainers need to make these kinds of changes. We currently have three issue that need a drift hash versioning improvement: 
+
+1. The team is looking to set cloud provider defaults for the: `nodeClassRef.apiVersion` and `nodeClassRef.kind` fields [#909](https://github.com/kubernetes-sigs/karpenter/issues/909)
+ -  **We will need to coordinate with the Azure team to make sure that `Kind` and `apiVersion` are also set on the (karpenter-provider-azure)[https://github.com/Azure/karpenter-provider-azure]**
+2. The aws team is blocked from fixing issue [#5447](https://github.com/aws/karpenter-provider-aws/issues/5447)
+3. The fix to the issue is adding volumeSize in the hashing mechanism of the `EC2NodeClass` (#5454)[https://github.com/aws/karpenter-provider-aws/pull/5454]
+
+However, this will cause all `NodeClaims` to be drifted as soon as a customer upgrades their Karpenter CRDs in their cluster. These two issues raise the need for Karpenter to understand different versions of the CRDs during a drift evaluation. The problem here is that a change on the Karpenter CRDs will cause all nodes on a customer’s cluster will be drifted without any changes to the node configuration.
+
+More generally, these are breaking changes to the hashing function that should not induce drift:
+
+1. A `NodePool` adds default values to an existing field
+2. A `NodePool` field is added with a default
+3. A `NodePool` field is removed
+
+## Solutions
+
+```
```

Comment:
This example spec feels out of place? Did you mean for this to be here?

Reply:
The example spec is meant for options 1 and option 2. Added a title to reflect that

Reply:
I personally wouldn't start with an example because you don't get the context for why that example is there. I'd generally give some background information for the solution and then lead that into the example spec. Happy to keep it as-is since we have been through a few iterations at this point, but that's just my two cents.

Reply:
I didn't want to duplicate the example, that was my major concern. Do you have any suggestion on a better way to structure the example?

Reply:
Personally, I would just put the example under Option 1 and then reference back to it with Option 2 if you need. That strikes me as a lot better than throwing it at the beginning without any context to start

Reply:
Updated to include the change